src/main/scala/io/iohk/ethereum/vm/ProgramContext.scala

```diff
@@ -1,13 +1,51 @@
 package io.iohk.ethereum.vm
 
-import akka.util.ByteString
+import io.iohk.ethereum.domain._
 
+
+object ProgramContext {
+  def apply(stx: SignedTransaction, blockHeader: BlockHeader, world: WorldStateProxy): ProgramContext = {
+    import stx.tx
+
+    val senderAddress = stx.recoveredSenderAddress.get // FIXME: get, it should be validated but...
+    val (world1, recipientAddress, program) = callOrCreate(world, tx, senderAddress)
+
+    val env = ExecEnv(recipientAddress, senderAddress, senderAddress, tx.gasPrice, tx.payload,
+      tx.value, program, blockHeader, callDepth = 0)
+
+    ProgramContext(env, tx.gasLimit, world1)
+  }
+
+  private def callOrCreate(world: WorldStateProxy, tx: Transaction, senderAddress: Address): (WorldStateProxy, Address, Program) = {
+    if (tx.receivingAddress.isEmpty) {
+      // contract create
+      val (address, world1) = world.newAddress(senderAddress)
+      val world2 = world1.saveAccount(address, Account.Empty)
+      val world3 = world2.transfer(senderAddress, address, tx.value)
+      val code = tx.payload
+
+      (world3, address, Program(code))
+
+    } else {
+      // message call
+      val world1 = world.transfer(senderAddress, tx.receivingAddress, tx.value)
+      val account = world1.getGuaranteedAccount(tx.receivingAddress)
+      val code = world1.getCode(account.codeHash)
+
+      (world1, tx.receivingAddress, Program(code))
+    }
+  }
+
+}
 /**
   * Input parameters to a program executed on the EVM. Apart from the code itself
   * it should have all (interfaces to) the data accessible from the EVM.
   *
   * @param env set of constants for the execution
   * @param startGas initial gas for the execution
-  * @param storage representation of the storage accociated with the contract to be executed
+  * @param world provides interactions with world state
   */
-case class ProgramContext(env: ExecEnv, startGas: BigInt, storage: Storage)
+case class ProgramContext(
+  env: ExecEnv,
+  startGas: BigInt, //TODO: should we move it to ExecEnv
+  world: WorldStateProxy)
```

Comment:
Now, this is quite important. Note `ProgramContext`, and consequently `ProgramState` and `ProgramResult`, are not type parameterised. This means that users of a `WorldStateProxy` implementation may need to cast it (`asInstanceOf`) to access methods which are not part of the interface, e.g. [in evm-runner](https://github.com/input-output-hk/etc-client/blob/demo/evmCallOpcodes/src/main/scala/io/iohk/ethereum/vmrunner/State.scala#L66)

However, type parameterising those classes also means we have to add type parameters to all the methods that use `ProgramState` in [OpCodes implementations](https://github.com/input-output-hk/etc-client/blob/feature/evmCallOpcodes/src/main/scala/io/iohk/ethereum/vm/OpCode.scala#L371), e.g:

```scala
protected def exec[WS <: WorldStateProxy](state: ProgramState[WS]): ProgramState[WS]
```

which produces some clutter.

Another way would be for the implementors to add new required methods to `WorldStateProxy`. But then that interface would no longer be unidirectional.

I ended up preferring casting (first time ever I think), which is quite safe in this case. But I'm interested in other opinions, especially @AlanVerbner 's since he's working on a implementation of `WorldStateProxy`.